Add a 'base_stat' attribute in [chance_to_hit]

newfrenchy83 opened the pull request:
I've already proposed this approach in https://github.com/wesnoth/wesnoth/pull/10279, but since creating a special [defense] weapon is a mistake, perhaps it will be accepted this time.

When this attribute is used, the resulting cth value is treated as a base value by chance_to_hit that doesn't use this attribute, other specials that affect chance to hit will  take precedence over this special.

stevecotton commented:
When you know that something has been discussed before, please link to the previous PR and comment on how the new version addresses the issues raised in the previous PR.

newfrenchy83 commented:
> When you know that something has been discussed before, please link to the previous PR and comment on how the new version addresses the issues raised in the previous PR.

discussed in beginning of https://github.com/wesnoth/wesnoth/pull/10279 PR

doofus-01 commented:
This PR seems to add more than what is mentioned in the title.  As far as the earlier discussion, what is one supposed to be looking for there?

newfrenchy83 commented:
> This PR seems to add more than what is mentioned in the title. As far as the earlier discussion, what is one supposed to be looking for there?

The two last commit only

newfrenchy83 commented:
> Please specify the full effects of this PR. Especially behavior changes must be mentioned.

when this attribute is used, the [chance_to_hit] behaves like an 'accuracy' or a 'parry' egulable that modifies the base value of cth, however a [chance_to_hit] will treat this value as a base value (if you have a [chance_to_hit]add=10 and base_cth=yes and a chance_to_hit of type marsksman, the latter will base on the base value +10 to decide if its value should replace the base or not, instead of having marskman 60 + 10, and this allows not to have to specify the base value to regulate the activity of add=10 for each [chance_to_hit] of type marskman likely to be encountered (especially if these belong to an extension and whose value cannot always be anticipated)

stevecotton commented:
Do I understand correctly, that chance to hit will be calculated in this order, with each step being able to add, subtract, multiply, divide and/or be cumulative with the result of the previous step?

1. [movetype][defense]
2. [abilities][defense]
3. [base_cth]
4. [chance_to_hit], [accuracy] and [parry], (I'm not sure which order these are applied)

The example in #10279, UtBS' "dazed" status, doesn't feel like a good example, "dazed" feels like a debuff that should be applied after the effects of marksman rather than before it.

newfrenchy83 commented:
> Do I understand correctly, that chance to hit will be calculated in this order, with each step being able to add, subtract, multiply, divide and/or be cumulative with the result of the previous step?
> 
> 1. [movetype][defense]
> 2. [abilities][defense]
> 3. [base_cth]
> 4. [chance_to_hit], [accuracy] and [parry], (I'm not sure which order these are applied)
> 


I changed the order for consistency.

1. [movetype][defense]
2. [abilities][defense]
3. accuracy= and parry=
4.  [chance_to_hit]base_cth=yes
5. default [chance_to_hit]

> The example in #10279, UtBS' "dazed" status, doesn't feel like a good example, "dazed" feels like a debuff that should be applied after the effects of marksman rather than before it.

the debuff for 'dazed' must be applied so that the value of marsksman remains at least 60 without this value being lowered to 50. currently the 'dazed' [chance_to_hit] must be disabled for 'magical' and 'marskman' and a second must be set to be active only when 'marskman' is used with a value greater than 70. And if a player plays a game of Utbs with an addon introducing units with [chance_to_hit] of type marsksman but with different values, 'dazed' will debuff these special weapons but spare 'marsksman', hence the idea of ​​'base_cth' to avoid this unequal treatment without indefinitely extending the list of [chance_to_hit] for this.

stevecotton commented:
> the debuff for 'dazed' must be applied so that the value of marsksman remains at least 60 without this value being lowered to 50.

Why? If 'dazed' makes it harder for units to hit, why should marksmen be immune to that debuff?

newfrenchy83 commented:
> Why? If 'dazed' makes it harder for units to hit, why should marksmen be immune to that debuff?

that's not the point, I made this change to preserve this behavior in a simpler and more universal way, if the person who programmed 'dazed' like this in the first place wants a change, let them know; I avoid introducing it deliberately.

newfrenchy83 commented:
@stevecotton don't forget this PR, please!

newfrenchy83 commented on data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/chance_to_hit/chance_to_hit_base_cth_sub_mixed_cumulative.cfg:
```diff
@@ -0,0 +1,44 @@
+##
+# Expected end state:
+# The side 1 leader's first weapon strikes 100 times(70-10)lower what value=100
+# The side 1 leader's second weapon strikes 100 times(70-10)lower what value=100
```
and added a test with "weapon strikes 90 times, because the normal [chance_to_hit]value,cumulative=70,yes is lower to 60 (base_value) + 30(base_cth [chance_to_hit]add=30"

stevecotton commented:
I'm still not sure about the feature, however I'm wondering whether a numerical `priority=` or `calculation_priority=` attribute could do both this and everything that `overwrite_specials` is used for. Although `overwrite_specials` can do more complex things, AFAICS it is only used for situations where "apply this value after all other calculations" would have the same result. My desire to reduce complexity would be satisfied if we added `calculation_priority=` but deprecated `overwrite_specials`, with a plan to remove it in 1.21 unless mainline or UMC showed a usecase that couldn't be done in a different way.

The name `base_cth` is specific to `chance_to_hit`, if another stat needed a similar feature then it would get a different name.

newfrenchy83 commented:
> > overwrite_special which is also used for [plague], [swarm] and [berserk]
> 
> Which UMC uses it? In mainline these seem to be only used in the unit tests.

I don't know but priority system can't be used here
